Mathlib/Topology/Category/Profinite/Nobeling.lean

```diff
@@ -0,0 +1,1549 @@
+    (IsLimit.ofIsoLimit (indexCone_isLimit hC) (Cones.ext (Iso.refl _) ?_))
+  · intro ⟨J⟩ ⟨K⟩ ⟨⟨⟨f⟩⟩⟩
+    ext x
+    have : iso_map C (· ∈ K) ∘ ProjRestricts C f = IndexFunctor.map C f ∘ iso_map C (· ∈ J) := by
+      ext _ i; exact dif_pos i.prop
+    exact congr_fun this x
+  · intro ⟨J⟩
+    ext x
+    have : iso_map C (· ∈ J) ∘ ProjRestrict C (· ∈ J) = IndexFunctor.π_app C (· ∈ J) := by
+      ext _ i; exact dif_pos i.prop
+    erw [← this]
+    rfl
+
+end Projections
+
+section Products
+
+/-- The locally constant map from `C` to `ℤ` whose `i`-th projection is given by 1 if
+    `f.val i = true`, and 0 otherwise.  -/
+def e (i : I) : LocallyConstant C ℤ where
+  toFun := fun f ↦ (if f.val i then 1 else 0)
+  isLocallyConstant := by
+    rw [IsLocallyConstant.iff_continuous]
+    exact (continuous_of_discreteTopology (f := fun (a : Bool) ↦ (if a then (1 : ℤ) else 0))).comp
+      ((continuous_apply i).comp continuous_subtype_val)
+
+/-- Formal products of the form `e C i₁ ··· e C iᵣ` with `i₁ > ··· > iᵣ`. -/
+def Products (I : Type*) [LinearOrder I] := {l : List I // l.Chain' (·>·)}
+
+namespace Products
+
+noncomputable
+instance : LinearOrder (Products I) :=
+  inferInstanceAs (LinearOrder {l : List I // l.Chain' (·>·)})
+
+@[simp]
+theorem lt_iff_lex_lt (l m : Products I) : l < m ↔ List.Lex (·<·) l.val m.val := by
+  cases l; cases m; rw [Subtype.mk_lt_mk]; exact Iff.rfl
+
+/-- The evaluation of of a formal product. -/
+def eval (l : Products I) := (l.1.map (e C)).prod
+
+/-- The predicate on products defining the basis. -/
+def isGood (l : Products I) : Prop :=
+  l.eval C ∉ Submodule.span ℤ ((Products.eval C) '' {m | m < l})
+
+theorem rel_head!_of_mem {i : I} {l : Products I} (hi : i ∈ l.val) : i ≤ l.val.head! :=
+  List.Sorted.le_head! (List.chain'_iff_pairwise.mp l.prop) hi
+
+theorem head!_le_of_lt {q l : Products I} (h : q < l) (hq : q.val ≠ []) :
+    q.val.head! ≤ l.val.head! :=
+  List.head!_le_of_lt l.val q.val h hq
+
+end Products
+
+/-- The type of good products. -/
+def GoodProducts := {l : Products I | l.isGood C}
+
+namespace GoodProducts
+
+/-- Evaluation of good products. -/
+def eval (l : {l : Products I // l.isGood C}) : LocallyConstant C ℤ :=
+  Products.eval C l.1
+
+theorem injective : Function.Injective (eval C) := by
+  intro ⟨a, ha⟩ ⟨b, hb⟩ h
+  dsimp [eval] at h
+  rcases lt_trichotomy a b with (h'|rfl|h')
+  · exfalso; apply hb; rw [← h]
+    exact Submodule.subset_span ⟨a, ⟨h',rfl⟩⟩
+  · rfl
+  · exfalso; apply ha; rw [h]
+    exact Submodule.subset_span ⟨b, ⟨h',rfl⟩⟩
+
+/-- The image of the good products in the module `LocallyConstant C ℤ`. -/
+def range := Set.range (GoodProducts.eval C)
+
+/-- The type of good products is equivalent to its image. -/
+noncomputable
+def equiv_range : GoodProducts C ≃ range C :=
+  Equiv.ofInjective (eval C) (injective C)
+
+theorem equiv_toFun_eq_eval : (equiv_range C).toFun = Set.rangeFactorization (eval C) := by rfl
```

Comment:
```suggestion
theorem equiv_toFun_eq_eval : (equiv_range C).toFun = Set.rangeFactorization (eval C) := rfl
```
Because not going into tactic mode saves vital milliseconds :-)